database.class.php

```diff
@@ -0,0 +1,287 @@
+<?php 
+require_once("./paginator.class.php");
+/*
+ * Database Class
+ * This class is used for database related (connect, insert, update, and delete) operations
+ * with PHP Data Objects (PDO)
+ */
+
+class Database{
+
+    /* 
+     * The variables have been declared as private. This
+     * means that they will only be available with the 
+     * Database class
+     */
+
+    private $db_driver;
+    private $db_host;
+    private $db_user;
+    private $db_pass;
+    private $db_name;
+    private $conn; // Hold the connection if it is active
+
+    public function __construct($db_driver, $db_name, $db_host = "localhost", $db_user = "root", $db_pass = "") {
+        // Setting values to local variables
+        $this->db_driver = $db_driver;
+        $this->db_host = $db_host;
+        $this->db_user = $db_user;
+        $this->db_pass = $db_pass;
+        $this->db_name = $db_name;
+
+        // Connect to the database
+        try {
+            $this->conn = new PDO($this->db_driver.":host=".$this->db_host.";dbname=".$this->db_name,$this->db_user,$this->db_pass);
+            $this->conn->setAttribute(PDO::ATTR_EMULATE_PREPARES, false);
+            $this->conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
+        } catch (PDOException $e) {
+            die($e->getMessage());
+        }
+
+    }
+
+    /*
+     * Function to validate data
+     * @param string the data to validate
+     */
+    private function dataVal($data){
+        $data = trim($data);
+        $data = stripslashes($data);
+        $data = htmlspecialchars($data);
+        return $data;
+    }
+
+    /*
+     * Function to bind values to prepared sql queries. This implementation helps with SQL Injection
+     * @param array the table with the data to bind in the statment
+     * @param string prepared sql query statement
+     */
+    protected function bind($table,$sql){
+       try {
+            $query = $this->conn->prepare($sql);
+            foreach($table as $key=>$val){
+                // Call dataVal function in order to validate data for SQL Injection
+                $query->bindValue(':'.$key,$this->dataVal($val));
+            }
+            // Return the prepared statment
+            return $query;
+        } catch (Exception $e) {
+            print_r($e->getMessage());
+            throw new Exception($e->getMessage());
+        }
+        
+    }
+
+    /*
+     * Function that executes and commits SQL queries with error handling.
+     * @param string prepared sql query statement with nameholders filled with values.
+     */
+    protected function execTransaction($query){
+        try {                
+                $this->conn->beginTransaction();
+                $transaction_exec = $query->execute();
+                $this->conn->commit();
+                return $transaction_exec;
```

Comment:
@abrakoup What's the meaning of start a transaction just to execute a single query and commit/rollback straight after?

Reply:
@tavantzo You are right, I should have placed the statement, at least, in each CRUD function or preferable in the bind function.

Reply:
@abrakoup : How it would be more useful inside the bind method body?

FYI: A common usage of db transactions goes like this....

```sql
START TRANSACTION;

INSERT INTO foo .....;
INSERT INTO bar .....;
UPDATE foo SET bar_id=@@IDENTITY WHERE .....
DELETE FROM `whaterver` WHERE.....

COMMIT/ROLLBACK
```
So basically transactions are meaningful in cases that you have to execute multiple sql statements in a row and that statement are somehow related e.g there is a foreign key etc.

So IHMO, a proper implementation would be somethind like

```php
$db->startTransaction();

$db->insert(...);
$db->insert(...);
....
$db->insert(...);

if ($allGood) {
$db->commitTransaction();
} else {
$db->rollbaclTransaction();
}

```